### ISSUE TITLE
Output from stdout for `docker exec` sometimes isn't returned / displayed

### ISSUE THREAD

**Issue by ncdc** [docker/docker#27289]
**Description**

Output from `docker exec` sometimes isn't returned/displayed. I've seen this in both 1.11.2 and 1.12.1. I *think* this may apply to `docker attach` as well, but I have only ever tested with exec.

**Steps to reproduce the issue:**
1. `sudo docker run -d --name nginx nginx`
2. `while true; do output=$(sudo docker exec nginx echo running in container); echo "${output}"; if [ "${output}" != "running in container" ]; then echo OMG; break; fi; done`

**Describe the results you received:**
After numerous iterations, each printing out `running in container` correctly, I see a blank line followed by my error indicator (`OMG`). Here is the tail of the results:

```
running in container
running in container
running in container

OMG
```

**Describe the results you expected:**

An endless display of `running in container` and never any blank lines.

**Additional information you deem important (e.g. issue happens only occasionally):**

This happens sporadically. I've seen it take anywhere from 10 minutes to about an hour or more to reproduce.

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-8:1-256115-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 250.8 MB
 Data Space Total: 107.4 GB
 Data Space Available: 8.914 GB
 Metadata Space Used: 876.5 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.110 (2015-10-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-38-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 992.3 MiB
Name: ubuntu-xenial
ID: 7D6Z:FC5P:U5BW:W622:Z4H2:PUZE:VDCV:NW4I:HGKD:DYUB:6XQO:VGEX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 19
 Goroutines: 29
 System Time: 2016-10-11T15:22:07.361345248Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

I've tested this in GCE using the GCI image with 1.11.2. I've also tested on VirtualBox using Ubuntu Xenial with both 1.11.2 and 1.12.1.

cc @runcom

**Comment by thaJeztah** [comment]
ping @mlaventure @cpuguy83 ptal

**Comment by runcom** [comment]
@ncdc could you try out with the latest master code?

**Comment by ncdc** [comment]
@runcom yes, but not today. Although it would be ideal to identify the issue and have a unit/int test for it if possible, because otherwise I'll just be running master in a loop waiting to see if it flakes.

**Comment by runcom** [comment]
alright, I remember we had a similar issue in 1.10.3 (w/o containerd+runc) and that has been fixed with https://github.com/docker/docker/pull/21018 - we're likely hitting the same scenario afaict (somehow)

**Comment by mlaventure** [comment]
The call to the `exec/.*/start` endpoint waits for all the streams to be closed (https://github.com/docker/docker/blob/9db46989294621ca63b200a690921ceb62c63db8/container/container.go#L460), so I'm a bit perplexed on how this can happen if the exec doesn't get killed or cancelled atm

**Comment by ncdc** [comment]
@runcom ... maybe? Note that this reproducer doesn't use a tty.

@mlarcher agreed, it is perplexing. Maybe the issue isn't in docker's exec flow but is in containerd or runc? I enabled debug logging for the daemon and the output was identical as best I could tell for both the "good" and "bad" exec cases. In other words, both showed `attach: stdout: begin` and `attach: stdout: end` (and the same for `stderr`).

**Comment by mlaventure** [comment]
It seems that `containerd`/`container-shim` are correctly waiting too.

The shim waits for the streams [here](https://github.com/docker/containerd/blob/0ac3cd1be170d180b2baed755e8f0da547ceb267/containerd-shim/main.go#L119) and the daemon waits for the shim to exit [here](https://github.com/docker/containerd/blob/0ac3cd1be170d180b2baed755e8f0da547ceb267/supervisor/exit.go#L76).

The pipe files are removed first [here](https://github.com/docker/containerd/blob/0ac3cd1be170d180b2baed755e8f0da547ceb267/supervisor/exit.go#L73) but that shouldn't affect anything.

I'll try to see if I can reproduce it locally

**Comment by saad-ali** [comment]
@mlaventure Any updates?

**Comment by mlaventure** [comment]
@saad-ali none, couldn't reproducing this after having exec run for a full day on 1.12.1 or 1.12.2

Have you tried moving to 1.12.2?

**Comment by ncdc** [comment]
@mlaventure I just tested 1.12.1 and it easily reproduced for me fairly quickly (< 10 mins)

**Comment by tonistiigi** [comment]
@ncdc Working on it. Should be same problem that is causing #27011

**Comment by resouer** [comment]
I also hit this issue, @tonistiigi you are fixing it right?

**Comment by thaJeztah** [comment]
@resouer fix was merged, and will be in docker 1.13

**Comment by calebamiles** [comment]
@thaJeztah, is there any intention of backporting this to the 1.11.x or 1.12.x series releases? @saad-ali  or @ncdc might have more context but I don't know if it will be possible to bump to Docker 1.13 immediately and this bug is causing a fair amount of flakiness in CI
